vernac/library.ml

```diff
@@ -272,18 +273,22 @@ let intern_from_file lib_resolver dir =
        DirPath.print lsd.md_name ++ spc () ++ str "and not library" ++
        spc() ++ DirPath.print dir ++ str ".");
   Feedback.feedback (Feedback.FileLoaded(DirPath.to_string dir, f));
-  lsd, lmd, digest_lmd, univs, digest_u, del_opaque
+  mk_intern_library lsd lmd digest_lmd univs digest_u del_opaque
 
+(* Returns the digest of a library, checks both caches to see what is loaded *)
 let rec intern_library ~intern (needed, contents as acc) dir =
   (* Look if in the current logical environment *)
-  try (find_library dir).libsum_digests, acc
-  with Not_found ->
-  (* Look if already listed and consequently its dependencies too *)
-  try (DPmap.find dir contents).library_digests, acc
-  with Not_found ->
-  let lsd, lmd, digest_lmd, univs, digest_u, del_opaque = intern dir in
-  let m = mk_intern_library lsd lmd digest_lmd univs digest_u del_opaque in
-  m.library_digests, intern_library_deps ~intern acc dir m
+  match find_library dir with
+  | Some loaded_lib -> loaded_lib.libsum_digests, acc
+  | None ->
+    (* Look if already listed in the accumulator *)
+    match DPmap.find_opt dir contents with
+    | Some interned_lib ->
+      interned_lib.library_digests, acc
+    | None ->
+      (* We intern the library, and then intern the deps *)
+      let m = intern dir in
```

Comment:
I would move the `if not (DirPath.equal dir lsd.md_name)` (now `.library_name`) here since `intern` is not necessarily `intern_from_file`

Reply:
Indeed, I've done as you propose. However see my comment.

Reply:
I cannot post a comment where I wanted, but indeed see this code:
```ocaml
    user_err
      (str "The file " ++ str file ++ str " contains library" ++ spc () ++
       DirPath.print library_name ++ spc () ++ str "and not library" ++
       spc() ++ DirPath.print dir ++ str ".")
```
Indeed, it makes sense to move it, however we will need a way to find the provenance from the parametric `intern` function.

Reply:
I assume you'd be OK to do like that, right?